src/Raven.Server/Documents/ShardedHandlers/ShardedDocumentHandler.cs

```diff
@@ -124,16 +113,100 @@ public async Task Get()
+        {
+            var metadataOnly = GetBoolValueQueryString("metadataOnly", required: false) ?? false;
+            var includePaths = GetStringValuesQueryString("include", required: false);
+            var etag = GetStringFromHeaders("If-None-Match");
+
+            using (ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            {
+                var docs = await context.ReadForMemoryAsync(RequestBodyStream(), "docs");
+                if (docs.TryGet("Ids", out BlittableJsonReaderArray array) == false)
+                    ThrowRequiredPropertyNameInRequest("Ids");
+
+                var ids = new string[array.Length];
+                for (int i = 0; i < array.Length; i++)
+                {
+                    ids[i] = array.GetStringByIndex(i);
+                }
+
+                var idsStringValues = new StringValues(ids);
+
+                if (TrafficWatchManager.HasRegisteredClients) //TODO - sharding: do we need that here?
+                    AddStringToHttpContext(idsStringValues.ToString(), TrafficWatchChangeType.Documents);
+
                 await GetDocumentsAsync(ids, includePaths, etag, metadataOnly, context);
             }
         }
 
+        [RavenShardedAction("/databases/*/docs", "DELETE")]
+        public async Task Delete()
+        {
+            var id = GetQueryStringValueAndAssertIfSingleAndNotEmpty("id");
+            
+            using (ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            {
+                var index = ShardedContext.GetShardIndex(context, id);
+
+                var cmd = new ShardedCommand(this, Headers.IfMatch);
+                await ShardedContext.RequestExecutors[index].ExecuteAsync(cmd, context);
+            }
+
+            NoContentStatus();
+        }
+
+        [RavenShardedAction("/databases/*/docs/class", "GET")]
+        public async Task GenerateClassFromDocument()
+        {
+            var id = GetStringQueryString("id");
+            var lang = (GetStringQueryString("lang", required: false) ?? "csharp")
+                .Trim().ToLowerInvariant();
+
+            using (ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            {
+                var index = ShardedContext.GetShardIndex(context, id);
+
+                var cmd = new ShardedCommand(this, Headers.None);
+                await ShardedContext.RequestExecutors[index].ExecuteAsync(cmd, context);
+                var document = cmd.Result;
+                if (document == null)
+                {
+                    HttpContext.Response.StatusCode = (int)HttpStatusCode.NotFound;
+                    return;
+                }
+
+                switch (lang)
+                {
+                    case "csharp":
+                        break;
+                    default:
+                        throw new NotImplementedException($"Document code generator isn't implemented for {lang}");
```

Comment:
Shouldn't we just reflect the error back to the user that we got from the destination?